### PR TITLE
Test: UnitTest for module_tddft

### PR DESCRIPTION
**Pull request by lyb9812**
Add unit tests for module_tddft.

### CODE REVIEW

File: source/module_base/blacs_connector.h [hunk 1]
```diff
@@ -33,7 +33,8 @@ extern "C"
 	void Cblacs_gridmap(int* icontxt, int *usermap, int ldumap, int nprow, int npcol);
 		// Informational and Miscellaneous
 	void Cblacs_gridinfo(int icontxt, int* nprow, int *npcol, int *myprow, int *mypcol);
-	int  Cblacs_pnum(int icontxt, int prow, int pcol);
-	void Cblacs_pcoord(int icontxt, int pnum, int *prow, int *pcol);
+    void Cblacs_gridinit(int* icontxt, char* layout, int nprow, int npcol);
```
Comment (hongriTianqi) [inline source/module_base/blacs_connector.h:36]: Please add unit tests for these newly added blas connectors, too. Thanks!

**Review by hongriTianqi — CHANGES_REQUESTED** [review]

File: source/module_hamilt_lcao/module_tddft/test/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,13 @@
+remove_definitions(-D __MPI)
+
+AddTest(
+  TARGET tddft_test
+  LIBS ${math_libs} base device
+  SOURCES tddft_test.cpp upsi_test1.cpp upsi_test2.cpp upsi_test3.cpp propagator_test1.cpp propagator_test2.cpp propagator_test3.cpp middle_hamilt_test.cpp norm_psi_test.cpp bandenergy_test.cpp ../upsi.cpp ../propagator.cpp ../middle_hamilt.cpp ../norm_psi.cpp ../bandenergy.cpp
```
Comment (hongriTianqi) [inline source/module_hamilt_lcao/module_tddft/test/CMakeLists.txt]: I suggest to add each tests separately with a specific name to facilitate checking of the corresponding functionalities.

**Review by hongriTianqi — CHANGES_REQUESTED** [review]

File: source/module_hamilt_lcao/module_tddft/test/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,34 @@
+remove_definitions(-D __MPI)
+
+add_library(tddft_test_lib tddft_test.cpp)
+
+AddTest(
+  TARGET middle_hamilt_test
```
Comment (hongriTianqi) [inline source/module_hamilt_lcao/module_tddft/test/CMakeLists.txt]: It is suggested to add prefix "tddft" for all the tests' name.

**Review by hongriTianqi — CHANGES_REQUESTED** [review]

**Comment by Qianruipku** [comment]
The CI for testing now takes a total of 40 minutes. What is the reason that the test takes 5 more minutes than before?

**Comment by hongriTianqi** [comment]
> The CI for testing now takes a total of 40 minutes. What is the reason that the test takes 5 more minutes than before?

More specifically, the case tests take 5 more minutes than before (1359 vs 1054 seconds).

**Review by mohanchen — APPROVED** [review]

**Review by hongriTianqi — CHANGES_REQUESTED** [review]
Can you please analyze the reason why case tests take 5 minutes more in this PR? @lyb9812

**Comment by hongriTianqi** [comment]
> Can you please analyze the reason why case tests take 5 minutes more in this PR?

In addition, multiple case tests deliver warnings with large derivation from the original results. This is unacceptable:

截屏2023-06-07 13 52 49

@mohanchen 

截屏2023-06-07 13 50 00

**Comment by hongriTianqi** [comment]
> > Can you please analyze the reason why case tests take 5 minutes more in this PR?
> 
> In addition, multiple case tests deliver warnings with large derivation from the original results. This is unacceptable:
> 
> 截屏2023-06-07 13 52 49
> @mohanchen
> 
> 截屏2023-06-07 13 50 00

This ISSUE is induced by my own PR #2477. I will take care of this issue.

**Review by hongriTianqi — APPROVED** [review]

**Review by hongriTianqi — CHANGES_REQUESTED** [review]
Wait, there are more warnings in this PR than PR #2477. Please check if this is related to this PR.

For comparison, the warnings appeared in the following cases in PR #2477.

截屏2023-06-07 14 15 27

**Comment by lyb9812** [comment]
> Can you please analyze the reason why case tests take 5 minutes more in this PR? @lyb9812

The total time of tests change from 34m to 40m since PR#2477.

**Comment by lyb9812** [comment]
> Wait, there are more warnings in this PR than PR #2477. Please check if this is related to this PR.
> 
> For comparison, the warnings appeared in the following cases in PR #2477.
> 
> 截屏2023-06-07 14 15 27

The warnings are the same as PR #2604 (the last received PR). So it is not introduced by this PR. It seems start from PR #2527 .

**Comment by hongriTianqi** [comment]
can you please check the two test cases of 601, which are related to TDDFT code. Can you please solve the warnnings in these two case tests? Thanks! @lyb9812 

tt

**Comment by hongriTianqi** [comment]
The warnings do not start from PR #2527, I am still searching for it ....

**Comment by hongriTianqi** [comment]
This issue is not induced by PR #2477, I am still searching for it ...

**Comment by hongriTianqi** [comment]
> The CI for testing now takes a total of 40 minutes. What is the reason that the test takes 5 more minutes than before?

5 more minutes are taken by EXX case tests, which were introduced in PR #2478. @lyb9812 @Qianruipku

**Comment by lyb9812** [comment]
> can you please check the two test cases of 601, which are related to TDDFT code. Can you please solve the warnnings in these two case tests? Thanks! @lyb9812
> 
> tt

There are no changes related to tddft in PR #2503 . The warnings in these two cases may be due to the random phase of wavefunction. I will confirm it after I complete the function of restart.

**Review by hongriTianqi — APPROVED** [review]

**Review by dyzheng — APPROVED** [review]

**Pull request merged**